### PR TITLE
Maybe fix notification tab hang/crash

### PR DESCRIPTION
**Pull request by mplorentz**
## Issues covered
Maybe #970, #1077 and maybe even #474

## Description
While debugging memory usage on an account that follows a lot of users I finally experienced the frequently reported notification tab hang/crash (#970). I caught it in the debugger and saw that the the main thread was deadlocked waiting on the `PersistenceController.backgroundContext` to save because a background thread was also accessing the background context in `NosNotification.markAllAsRead(...)`. The problem is that `PersistenceController.saveAll()` was not saving the `backgroundContext` on the the `backgroundContext`'s associated thread using `backgroundContext.perform {...}`.

I think this deadlock in a production build would look like a hang or a crash - maybe even a crash that doesn't capture a stack trace. If this crash doesn't capture a stack trace then maybe the crash reports we thought were for out of memory errors were really this issue. We should be able to tell if this is true after #539 is deployed to production.

## How to test
The problem was difficult to reproduce but I was able to do it a couple times by backgrounding/foregrounding the app a lot and tapping between the notification and home tabs (which triggers the `NotificationsView` to mark everything as read).

### CODE REVIEW

File: Nos/Models/CoreData/Event+CoreDataClass.swift [hunk 1]
```diff
@@ -1042,7 +1042,7 @@ public class Event: NosManagedObject, VerifiableEvent {
             
             event.isRead = true
             do {
-                try context.save()
+                try context.saveIfNeeded()
```
Comment (mplorentz) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:1045]: I noticed some unnecessary saves happening here, so I fixed them.

File: Nos/NosApp.swift [hunk 2]
```diff
@@ -33,15 +33,19 @@ struct NosApp: App {
...
-                    } else if newPhase == .active {
+                    case .active:
                         Log.info("Scene change: active")
-                    } else if newPhase == .background {
+                    case .background:
                         Log.info("Scene change: background")
-                        try? persistenceController.saveAll()
+                        Task {
+                            // TODO: save all contexts, not just the view and background.
+                            try await persistenceController.saveAll()
```
Comment (mplorentz) [inline Nos/NosApp.swift:45]: The app always passed through "inactive" before going to the background, so we were calling `saveAll()` too much. The docs say that the app will always go into the background before being closed, so it seems safe to only save there.

**Review by joshuatbrown — APPROVED** [review]
This looks great, and I can't get it to hang! Awesome!

**Comment by joshuatbrown** [comment]
Oh, and can I also just say that Core Data is confusing?

**Comment by mplorentz** [comment]
> Oh, and can I also just say that Core Data is confusing?

Yes, let me call this out as a PSA @joshuatbrown @martindsq:
_Do not .save() an NSManagedObjectContext except inside a context.perform { } block. It could deadlock the app._

_... unless it is the `viewContext` which is bound to the main thread. In that case mark the function that calls save @MainActor._

_... or unless you call .saveIfNeeded() because that does its own call to performAndWait { }._

Clear as mud?

I went through and did an audit of our code and I don't see any places we are breaking these rules. But this is another Core Data rule that the compiler will not help us with.

**Comment by joshuatbrown** [comment]
Can't they add some compiler checks for us? That would make my life easier, and I'd like my life to be easier.

**Comment by mplorentz** [comment]
@joshuatbrown these are the sorts of checks that Swift Structured Concurrency is supposed to enable, but it's unclear how much support Core Data (or even SwiftData) will have for them. The fact that there are no videos from WWDC 2024 about concurrency and Core Data or SwiftData has me worried 😭

**Pull request merged**